cmd/analytics/main.go

```diff
@@ -0,0 +1,168 @@
+
+	// Histograms for Storage Tries.
+	histStorageTrieDepths := histogram.New[int]("Storage Trie - Depths")
+	histStorageTriesNumSlots := histogram.New[int64]("Storage Trie - Number of used slots")
+
+	iter := t.NodeIterator(nil)
+	for iter.Next(true) {
+		if iter.Leaf() {
+			leafNodes++
+
+			// State Trie analysis.
+			pathNodeTypes := iter.Stack()
+			histStateTrieDepths.Observe(len(pathNodeTypes) - 1)
+			histStatePathTypes.Observe(toShortPathTypes(pathNodeTypes))
+
+			// Storage tries analysis.
+			var acc types.StateAccount
+			if err := rlp.DecodeBytes(iter.LeafBlob(), &acc); err != nil {
+				log.Fatalf("invalid account encountered during traversal: %s", err)
+			}
+			if acc.Root != emptyRoot {
+				storageTries++
+				id := trie.StorageTrieID(trieRoot, common.BytesToHash(iter.LeafKey()), acc.Root)
+				storageTrie, err := trie.NewStateTrie(id, triedb)
+				if err != nil {
+					log.Fatalf("failed to open storage trie: %s", err)
+				}
+
+				var storageTriesNumSlots, storageSlotCumDepth int64
+				storageIter := storageTrie.NodeIterator(nil)
+				for storageIter.Next(true) {
+					if storageIter.Leaf() {
+						if ctx.Err() != nil {
+							break
+						}
+						storageTriesNumSlots += 1
+						pathNodeTypes := storageIter.Stack()
+						storageSlotCumDepth += int64(len(pathNodeTypes) - 1)
+					}
+				}
+				histStorageTrieDepths.Observe(int(storageSlotCumDepth / storageTriesNumSlots))
+				histStorageTriesNumSlots.Observe(storageTriesNumSlots)
+
+				if storageIter.Error() != nil {
+					log.Fatalf("Failed to traverse storage trie: %s", err)
+				}
+			}
+		}
+
+		if time.Since(lastReport) > time.Minute*15 {
+			// State Trie stdout reports.
+			fmt.Printf("Walked %d (EOA + SC) accounts:\n", leafNodes)
+			histStateTrieDepths.Print(os.Stdout)
+			histStatePathTypes.Print(os.Stdout)
+			fmt.Println()
+
+			// Storage tries stdout reports.
+			fmt.Printf("Walked %d Storage Tries:\n", storageTries)
+			histStorageTrieDepths.Print(os.Stdout)
+			histStorageTriesNumSlots.Print(os.Stdout)
+
+			fmt.Printf("-----\n\n")
+
+			lastReport = time.Now()
+
+			// Persist .csv.
+			// State Trie.
+			histStateTrieDepths.ToCSV("statetrie_depth.csv")
+			histStatePathTypes.ToCSV("statetrie_pathtypes.csv")
+
+			// Storage Tries.
+			histStorageTrieDepths.ToCSV("storagetrie_depth.csv")
+			histStorageTriesNumSlots.ToCSV("storagetrie_numslots.csv")
+
+		}
+
+		if ctx.Err() != nil {
+			return
+		}
+	}
+	if iter.Error() != nil {
+		log.Fatalf("iterating trie: %s", iter.Error())
+	}
+}
+
+func toShortPathTypes(nodeTypes []string) string {
+	var sb strings.Builder
+	for i, nodeType := range nodeTypes[:len(nodeTypes)-1] {
+		switch nodeType {
+		case "*trie.shortNode":
+			if i == len(nodeTypes)-2 {
+				sb.WriteString("L.")
+				continue
+			}
+			sb.WriteString("E.")
+		case "*trie.fullNode":
+			sb.WriteString("B.")
+		default:
+			panic("unkown node type")
+		}
+	}
+	strPath := sb.String()
+
+	return strPath[:len(strPath)-1]
+}
```

Comment:
This method does a transformation of a `[]string` which has reflection types of the underlying MPT nodes:
- `trie.shortNode`
- `trie.fullNode`
And transforms that to a pretty string like `B.B.E.B.L` (Branch -> Branch -> Extension -> Branch -> Leaf).

There're two somewhat opaque important details about this.

First, the `nodeTypes[len(nodeTypes)-1]` should be ignored since this is a `trie.valueNode`. The reality is that [this type](https://github.com/ethereum/go-ethereum/blob/master/trie/node.go#L47) isn't exactly a "logical node" but just the raw value of the leaf. (See the next paragraph to make more sense about this).

The `*trie.shortNode` case has an important detail related to the internals of MPT in Geth.
The `trie.shortNode` type represents both extension nodes and leaf nodes, depending on what the underlying [(trie.shortNode).Value](https://github.com/ethereum/go-ethereum/blob/master/trie/node.go#L43) is. If that `Value` is a fullNode it means its an Extension Node, if it's a [valueNode](https://github.com/ethereum/go-ethereum/blob/master/trie/node.go#L47) it's a leaf.

If we didn't account for what was explained in the last two paragraphs, every branch would end in (... -> Extension -> Leaf), and that doesn't make sense "logically". It's true that in Geth this is how branches look like, but if we imagine the MPT conceptually having a (... -> Extension -> Leaf) ending doesn't make sense.

This isn't a bug or anything like that. It's just a consequence of leaking internal details on how Geth represents things, which probably have a good reason to solve things more easily while working with MPTs.